### PR TITLE
Generalize and improve re-uploading of images to Cloudinary

### PR DESCRIPTION
**Pull request by jimrandomh**
Long ago, I wrote a manually-runnable script to scan a post for `` tags, download those images, re-upload them on Cloudinary and edit the post to point at the CDN'd versions. Recently in https://github.com/ForumMagnum/ForumMagnum/pull/6038 Will fixed this up a bit and made it auto-run on post edit/submit.

This fixes up some remaining issues in the reupload script. In particular:
 * We guard against duplicate uploads (before, each re-edit of a post produced a duplicate image upload, potentially heavily bloating our Cloudinary acount)
 * It applies to non-post editables like tags, user bios, etc
 * We translate the `srcset` attribute as well as the `src` attribute.

Fixes https://github.com/ForumMagnum/ForumMagnum/issues/6132

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203403442055279) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by s-cheng** [comment]
@jimrandomh FYI I tested this with posts locally, and noticed that we were uploading each image 3 times (both to Cloudinary and to the Images collection). Looks like it's because the `editAsync` hook that calls `convertImagesInObject` gets added once per editable field in the collection, so Posts gets it 3 times.

**Comment by jimrandomh** [comment]
Oh, hmm. That does seem important to fix. There's handling to prevent duplicate uploads, but it looks like there's a race condition there with uploads that start simultaneously.

**Comment by jimrandomh** [comment]
Fixed duplicate uploads from the multiple editable fields. There's still some duplication caused by images having multiple DPI levels; CkEditor populates the `srcset` field with a collection of different zoom levels. I think this is worth fixing eventually, but not a blocker to merging this now.

**Comment by jpaddison3** [comment]
@jimrandomh did you mean to push code?

**Comment by jimrandomh** [comment]
I meant to put this back in reviewable state but didn't change the assignment.

**Comment by jpaddison3** [comment]
@jimrandomh you say you

> Fixed duplicate uploads from the multiple editable fields.

however I don't see any commits in which you push any fixes?

**Comment by jimrandomh** [comment]
Whoops, there was indeed an unpushed commit. Should be pushed now.

File: packages/lesswrong/server/scripts/convertImagesToCloudinary.ts [hunk 1]
```diff
@@ -67,58 +86,124 @@ async function convertImagesInHTML(html: string, originDocumentId: string, urlFi
...
+ * @param collectionName The collection that this object is in
+ * @param _id - The object to reupload images for
  * @param urlFilterFn - A function that takes a URL and returns true if it should be mirrored, by default all URLs are mirrored except those in getImageUrlWhitelist()
  * @returns The number of images that were mirrored
  */
-export async function convertImagesInPost(postId: string, urlFilterFn: (url: string) => boolean = () => true): Promise<number> {
+export async function convertImagesInObject(
+  collectionName: CollectionNameString,
+  _id: string,
+  fieldName = "contents",
```
Comment (s-cheng) [inline packages/lesswrong/server/scripts/convertImagesToCloudinary.ts:180]: Could you add this to the docstring?

File: packages/lesswrong/server/scripts/convertImagesToCloudinary.ts [hunk 2]
```diff
@@ -67,58 +86,124 @@ async function convertImagesInHTML(html: string, originDocumentId: string, urlFi
...
     
-    const newVersion = await getNextVersion(postId, "patch", "contents", false);
+    const newVersion = await getNextVersion(_id, "patch", fieldName, false);
     const now = new Date();
     // NOTE: we use the post contents rather than the revision contents because we don't
     // create a revision for no-op edits (this is arguably a bug)
-    const oldHtml = post.contents.html;
-    const {count: uploadCount, html: newHtml} = await convertImagesInHTML(oldHtml, postId, urlFilterFn);
+    const oldHtml = obj[fieldName].html;
+    const {count: uploadCount, html: newHtml} = await convertImagesInHTML(oldHtml, _id, urlFilterFn);
```
Comment (s-cheng) [inline packages/lesswrong/server/scripts/convertImagesToCloudinary.ts:206]: The `Converted ${uploadCount} images` message below isn't accurate anymore - would be nice to make that message clearer and/or more useful

**Review by s-cheng — APPROVED** [review]
Looks good, thanks!

**Pull request merged**